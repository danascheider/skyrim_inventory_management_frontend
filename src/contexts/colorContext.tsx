import { createContext, ReactElement } from 'react'
import { YELLOW, type ColorScheme } from '../utils/colorSchemes'

const ColorContext = createContext<ColorScheme>(YELLOW)

interface ProviderProps {
  colorScheme: ColorScheme
  children: ReactElement
}

const ColorProvider = ({ colorScheme, children }: ProviderProps) => (
  <ColorContext.Provider value={colorScheme}>{children}</ColorContext.Provider>
)

export { ColorProvider, ColorContext }
