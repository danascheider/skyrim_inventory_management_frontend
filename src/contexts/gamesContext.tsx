import { createContext, useCallback, useEffect, useState } from 'react'
import { signOutWithGoogle } from '../firebase'
import { type Game } from '../types/games'
import { type ProviderProps } from '../types/contexts'
import { useGoogleLogin, usePageContext } from '../hooks/contexts'
import { ApiError } from '../types/errors'
import { LOADING, DONE, ERROR, type LoadingState } from '../utils/loadingStates'
import { getGames, deleteGame } from '../utils/simApi'

// TODO: Add default values for context provider as follows:
//       const createGame = (game: Game) => {
//                                            postGames(game: Game, token: '')
//                                              .then((status: number, json: string) => /* noop */)
//                                          }
//       This will be useful in Storybook.

const NOT_FOUND_MESSAGE =
  "Oops! We couldn't find the game you're looking for. Please refresh and try again."
const UNEXPECTED_ERROR_MESSAGE =
  "Oops! Something unexpected went wrong. We're sorry! Please try again later."

export interface GamesContextType {
  games: Game[]
  gamesLoadingState: LoadingState
  // createGame: (game: Game) => void
  // updateGame: (gameId: number, attrs: Game) => void
  destroyGame: (gameId: number) => void
}

export const GamesContext = createContext<GamesContextType>({
  games: [],
  gamesLoadingState: LOADING,
  destroyGame: () => {},
})

export const GamesProvider = ({ children }: ProviderProps) => {
  const { user, token, authLoading, requireLogin } = useGoogleLogin()
  const [gamesLoadingState, setGamesLoadingState] = useState(LOADING)
  const [games, setGames] = useState<Game[]>([])
  const { setFlashProps } = usePageContext()

  /*
   *
   * Retrieve all the current user's games from the API
   *
   */

  const fetchGames = useCallback(() => {
    if (user && token) {
      getGames(token)
        .then(({ json }) => {
          if (Array.isArray(json)) {
            setGames(json)
            setGamesLoadingState(DONE)
          }
        })
        .catch((e: ApiError) => {
          if (import.meta.env.DEV)
            console.error(`Error ${e.name}: ${e.message}`)

          if (e.code === 401) {
            signOutWithGoogle()
          } else {
            setGames([])
            setGamesLoadingState(ERROR)
            setFlashProps({
              type: 'error',
              message: e.message,
              hidden: false,
            })
          }
        })
    }
  }, [user, token])

  /*
   *
   * Destroy the requested game and update the `games` array
   *
   */

  const destroyGame = useCallback(
    (gameId: number) => {
      if (user && token) {
        deleteGame(gameId, token)
          .then(({ status }) => {
            if (status === 204) {
              const newGames = games.filter(({ id }) => id !== gameId)
              setGames(newGames)
              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your game has been deleted.',
              })
            }
          })
          .catch((e: ApiError) => {
            if (import.meta.env.DEV)
              console.log(`Error ${e.name}: ${e.message}`)

            const message =
              e.code === 404 ? NOT_FOUND_MESSAGE : UNEXPECTED_ERROR_MESSAGE

            setFlashProps({
              hidden: false,
              type: 'error',
              message,
            })
          })
      }
    },
    [user, token, games]
  )

  const value = {
    games,
    gamesLoadingState,
    destroyGame,
  }

  useEffect(() => {
    requireLogin()
  }, [requireLogin])

  useEffect(() => {
    if (authLoading) return

    fetchGames()
  }, [authLoading, fetchGames])

  return <GamesContext.Provider value={value}>{children}</GamesContext.Provider>
}
