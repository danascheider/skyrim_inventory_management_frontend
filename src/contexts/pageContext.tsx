import { createContext, useState, useEffect, useRef } from 'react'
import { type ApiCalls, type Resource, type HttpVerb } from '../types/apiCalls'
import { ProviderProps } from '../types/contexts'
import { type FlashProps, type ModalProps } from '../types/pageContext'

interface PageContextType {
  flashProps: FlashProps
  setFlashProps: (props: FlashProps) => void
  modalProps: ModalProps
  setModalProps: (props: ModalProps) => void
  apiCallsInProgress: ApiCalls
  addApiCall: (key: Resource, value: HttpVerb) => void
  removeApiCall: (key: Resource, value: HttpVerb) => void
}

const defaultFlashProps: FlashProps = {
  type: 'info',
  message: '',
  hidden: true,
}

const defaultModalProps: ModalProps = {
  hidden: true,
  children: <></>,
}

const defaultApiCallStatus = {
  games: [],
  shoppingLists: [],
  shoppingListItems: [],
}

const PageContext = createContext<PageContextType>({
  flashProps: defaultFlashProps,
  modalProps: defaultModalProps,
  apiCallsInProgress: defaultApiCallStatus,
  setFlashProps: () => {},
  setModalProps: () => {},
  addApiCall: () => {},
  removeApiCall: () => {},
})

const PageProvider = ({ children }: ProviderProps) => {
  const [flashProps, setFlashProps] = useState<FlashProps>(defaultFlashProps)
  const [modalProps, setModalProps] = useState<ModalProps>(defaultModalProps)
  const [apiCallsInProgress, SetApiCallsInProgress] =
    useState<ApiCalls>(defaultApiCallStatus)

  const flashVisibleSince = useRef(0)

  const addApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: [...apiCallsInProgress[key], value],
    })
  }

  const removeApiCall = (key: Resource, value: HttpVerb) => {
    SetApiCallsInProgress({
      ...apiCallsInProgress,
      [key]: apiCallsInProgress[key].filter((verb) => verb !== value),
    })
  }

  const value = {
    flashProps,
    modalProps,
    apiCallsInProgress,
    setFlashProps,
    setModalProps,
    addApiCall,
    removeApiCall,
  }

  useEffect(() => {
    if (flashProps.hidden === true) return

    flashVisibleSince.current = Number(new Date())

    setTimeout(() => {
      if (Number(new Date()) - flashVisibleSince.current >= 4000) {
        setFlashProps({ ...flashProps, hidden: true })
        flashVisibleSince.current = 0
      }
    }, 4000)
  }, [flashProps])

  return <PageContext.Provider value={value}>{children}</PageContext.Provider>
}

export { PageContext, PageProvider }
