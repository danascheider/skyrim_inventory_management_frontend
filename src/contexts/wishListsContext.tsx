import { createContext, useEffect, useState, useRef, useCallback } from 'react'
import { type CallbackFunction } from '../types/functions'
import {
  type RequestWishListItem,
  type RequestWishList,
  type ResponseWishList,
} from '../types/apiData'
import { type ProviderProps } from '../types/contexts'
import { type ApiError } from '../types/errors'
import { LOADING, DONE, ERROR, type LoadingState } from '../utils/loadingStates'
import {
  postWishLists,
  getWishLists,
  patchWishList,
  deleteWishList,
  postWishListItems,
  patchWishListItem,
  deleteWishListItem,
} from '../utils/api/simApi'
import { useQueryString } from '../hooks/useQueryString'
import {
  useGoogleLogin,
  usePageContext,
  useGamesContext,
} from '../hooks/contexts'

const UNEXPECTED_ERROR_MESSAGE =
  "Oops! Something unexpected went wrong. We're sorry! Please try again later."

export interface WishListsContextType {
  wishLists: ResponseWishList[]
  wishListsLoadingState: LoadingState
  createWishList: (
    attributes: RequestWishList,
    onSuccess?: CallbackFunction | null,
    onError?: CallbackFunction | null,
    idToken?: string | null,
    retries?: number
  ) => void
  updateWishList: (
    listId: number,
    attributes: RequestWishList,
    onSuccess?: CallbackFunction | null,
    onError?: CallbackFunction | null,
    idToken?: string | null,
    retries?: number
  ) => void
  destroyWishList: (
    listId: number,
    onSuccess?: CallbackFunction | null,
    onError?: CallbackFunction | null,
    idToken?: string | null,
    retries?: number
  ) => void
  createWishListItem: (
    listId: number,
    attributes: RequestWishListItem,
    onSuccess?: CallbackFunction | null,
    onError?: CallbackFunction | null,
    idToken?: string | null,
    retries?: number
  ) => void
  updateWishListItem: (
    itemId: number,
    attributes: RequestWishListItem,
    onSuccess?: CallbackFunction | null,
    onError?: CallbackFunction | null,
    idToken?: string | null,
    retries?: number
  ) => void
  destroyWishListItem: (
    itemId: number,
    onSuccess?: CallbackFunction | null,
    onError?: CallbackFunction | null,
    idToken?: string | null,
    retries?: number
  ) => void
}

export const WishListsContext = createContext<WishListsContextType>({
  wishLists: [] as ResponseWishList[],
  wishListsLoadingState: LOADING,
  createWishList: () => {},
  updateWishList: () => {},
  destroyWishList: () => {},
  createWishListItem: () => {},
  updateWishListItem: () => {},
  destroyWishListItem: () => {},
})

export const WishListsProvider = ({ children }: ProviderProps) => {
  const { token, authLoading, requireLogin, withTokenRefresh, signOut } =
    useGoogleLogin()
  const { setFlashProps, addApiCall, removeApiCall } = usePageContext()
  const { gamesLoadingState, games } = useGamesContext()
  const queryString = useQueryString()
  const [activeGame, setActiveGame] = useState<number | null>(null)
  const [wishListsLoadingState, setWishListsLoadingState] =
    useState(LOADING)
  const [wishLists, setWishLists] = useState<ResponseWishList[]>([])
  const previousTokenRef = useRef(token)

  /**
   *
   * General handler for any ApiError
   *
   */

  const handleApiError = (e: ApiError, resource?: 'list' | 'list item') => {
    if (import.meta.env.DEV) console.error(e.message)

    if (e.code === 401) {
      signOut()
    } else if (e.code === 422) {
      setFlashProps({
        hidden: false,
        type: 'error',
        header: `${e.message.length} error(s) prevented your wish ${
          resource || 'list'
        } from being saved:`,
        message: e.message,
      })
    } else {
      const message =
        e.code === 405
          ? "You can't manually manage an aggregate list."
          : UNEXPECTED_ERROR_MESSAGE

      setFlashProps({
        hidden: false,
        type: 'error',
        message,
      })
    }
  }

  /**
   *
   * Create wish list for the active game
   *
   */

  const createWishList = useCallback(
    (
      attributes: RequestWishList,
      onSuccess?: CallbackFunction | null,
      onError?: CallbackFunction | null,
      idToken?: string | null,
      retries?: number
    ) => {
      if (!activeGame) {
        setFlashProps({
          hidden: false,
          type: 'warning',
          message:
            'You must select a game from the dropdown before creating a wish list.',
        })

        return
      }

      idToken ??= token

      if (idToken) {
        addApiCall('wishLists', 'post')
        postWishLists(activeGame, attributes, idToken)
          .then(({ json }) => {
            if (Array.isArray(json)) {
              if (json.length == 2) {
                setWishLists(json)
              } else {
                const newWishLists = [...wishLists]
                newWishLists.splice(1, 0, json[0])
                setWishLists(newWishLists)
              }

              removeApiCall('wishLists', 'post')

              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your wish list has been created.',
              })

              onSuccess && onSuccess()
            }
          })
          .catch((e: ApiError) => {
            retries ??= 1

            if (e.code === 401 && retries > 0) {
              return withTokenRefresh((newToken) => {
                createWishList(
                  attributes,
                  onSuccess,
                  onError,
                  newToken,
                  (retries as number) - 1
                )
              })
            } else if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "The game you've selected doesn't exist, or doesn't belong to you. Please select another game and try again.",
              })
            } else {
              handleApiError(e)
            }

            removeApiCall('wishLists', 'post')
            onError && onError()
          })
      }
    },
    [token, activeGame, wishLists]
  )

  /**
   *
   * Fetch wish lists for the active game and set
   * them as the wishLists array
   *
   */

  const setWishListsFromApi = (
    idToken: string | null = token,
    retries: number = 1
  ) => {
    if (!activeGame || !idToken) return

    addApiCall('wishLists', 'get')
    getWishLists(activeGame, idToken)
      .then(({ json }) => {
        if (Array.isArray(json)) {
          setWishLists(json)
          setWishListsLoadingState(DONE)
          removeApiCall('wishLists', 'get')
        }
      })
      .catch((e: ApiError) => {
        if (e.code === 401 && retries > 0) {
          return withTokenRefresh((newToken) => {
            setWishListsFromApi(newToken, retries - 1)
          })
        } else if (e.code === 404) {
          setFlashProps({
            hidden: false,
            type: 'error',
            message:
              "The game you've selected doesn't exist, or doesn't belong to you. Please select another game and try again.",
          })
        } else {
          handleApiError(e)
        }

        removeApiCall('wishLists', 'get')
        setWishLists([])
        setWishListsLoadingState(ERROR)
      })
  }

  const fetchWishLists = useCallback(() => {
    if (token && activeGame) {
      setWishListsLoadingState(LOADING)
      setWishListsFromApi()
    }
  }, [token, activeGame])

  /**
   *
   * Update specified wish list
   *
   */

  const updateWishList = useCallback(
    (
      listId: number,
      attributes: RequestWishList,
      onSuccess?: CallbackFunction | null,
      onError?: CallbackFunction | null,
      idToken?: string | null,
      retries?: number
    ) => {
      idToken ??= token

      if (idToken) {
        addApiCall('wishLists', 'patch')
        patchWishList(listId, attributes, idToken)
          .then(({ status, json }) => {
            if (status === 200) {
              const newWishLists = [...wishLists]
              const index = newWishLists.findIndex(
                ({ id }) => id === listId
              )
              newWishLists[index] = json

              setWishLists(newWishLists)
              removeApiCall('wishLists', 'patch')
              onSuccess && onSuccess()
            } else {
              // This won't happen but TypeScript doesn't know that
              removeApiCall('wishLists', 'patch')

              setFlashProps({
                hidden: false,
                type: 'error',
                message: UNEXPECTED_ERROR_MESSAGE,
              })

              onError && onError()
            }
          })
          .catch((e: ApiError) => {
            retries ??= 1

            if (e.code === 401 && retries > 0) {
              return withTokenRefresh((newToken) => {
                updateWishList(
                  listId,
                  attributes,
                  onSuccess,
                  onError,
                  newToken,
                  (retries as number) - 1
                )
              })
            } else if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "The wish list you tried to update doesn't exist, or doesn't belong to you. Please refresh and try again.",
              })
            } else {
              handleApiError(e)
            }

            removeApiCall('wishLists', 'patch')
            onError && onError()
          })
      }
    },
    [token, wishLists]
  )

  /**
   *
   * Destroy specified wish list
   *
   */

  const destroyWishList = useCallback(
    (
      listId: number,
      onSuccess?: CallbackFunction | null,
      onError?: CallbackFunction | null,
      idToken?: string | null,
      retries?: number
    ) => {
      idToken ??= token

      if (idToken) {
        addApiCall('wishLists', 'delete')
        deleteWishList(listId, idToken)
          .then(({ json }) => {
            if ('errors' in json) {
              // This case should never happen because normally an ApiError
              // will be thrown for any response that includes this key, but
              // TypeScript doesn't know that.
              removeApiCall('wishLists', 'delete')

              setFlashProps({
                hidden: false,
                type: 'error',
                message: UNEXPECTED_ERROR_MESSAGE,
              })

              onError && onError()
            } else {
              const newWishLists = wishLists

              if (json.aggregate) newWishLists[0] = json.aggregate

              for (const deletedId of json.deleted) {
                const index = newWishLists.findIndex(
                  (list) => list.id === deletedId
                )
                newWishLists.splice(index, 1)
              }

              setWishLists(newWishLists)
              removeApiCall('wishLists', 'delete')

              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your wish list has been deleted.',
              })

              onSuccess && onSuccess()
            }
          })
          .catch((e: ApiError) => {
            retries ??= 1

            if (e.code === 401 && retries > 0) {
              return withTokenRefresh((newToken) => {
                destroyWishList(
                  listId,
                  onSuccess,
                  onError,
                  newToken,
                  (retries as number) - 1
                )
              })
            } else if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "The wish list you tried to delete doesn't exist, or doesn't belong to you. Please refresh and try again.",
              })
            } else {
              handleApiError(e)
            }

            removeApiCall('wishLists', 'delete')
            onError && onError()
          })
      }
    },
    [token, wishLists]
  )

  /**
   *
   * Create a new wish list item
   *
   */

  const createWishListItem = useCallback(
    (
      listId: number,
      attributes: RequestWishListItem,
      onSuccess?: CallbackFunction | null,
      onError?: CallbackFunction | null,
      idToken?: string | null,
      retries?: number
    ) => {
      idToken ??= token

      if (idToken) {
        addApiCall('wishListItems', 'post')
        postWishListItems(listId, attributes, idToken)
          .then(({ status, json }) => {
            if (status === 200 || status === 201) {
              const newWishLists = [...wishLists]

              for (let list of json) {
                const index = newWishLists.findIndex(
                  ({ id }) => id === list.id
                )
                newWishLists[index] = list
              }

              setWishLists(newWishLists)

              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your wish list item has been created.',
              })

              onSuccess && onSuccess()
            } else {
              setFlashProps({
                hidden: false,
                type: 'error',
                message: UNEXPECTED_ERROR_MESSAGE,
              })

              onError && onError()
            }

            removeApiCall('wishListItems', 'post')
          })
          .catch((e: ApiError) => {
            retries ??= 1

            if (e.code === 401 && retries > 0) {
              return withTokenRefresh((newToken) => {
                createWishListItem(
                  listId,
                  attributes,
                  onSuccess,
                  onError,
                  newToken,
                  (retries as number) - 1
                )
              })
            } else if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "The wish list you tried to add an item to doesn't exist, or doesn't belong to you. Please refresh and try again.",
              })
            } else {
              handleApiError(e, 'list item')
            }

            removeApiCall('wishListItems', 'post')
            onError && onError()
          })
      }
    },
    [token, wishLists]
  )

  /**
   *
   * Update a wish list item
   *
   */

  const updateWishListItem = useCallback(
    (
      itemId: number,
      attributes: RequestWishListItem,
      onSuccess?: CallbackFunction | null,
      onError?: CallbackFunction | null,
      idToken?: string | null,
      retries?: number
    ) => {
      idToken ??= token

      if (idToken) {
        addApiCall('wishListItems', 'patch')

        patchWishListItem(itemId, attributes, idToken)
          .then(({ status, json }) => {
            if (status === 200) {
              const newWishLists = [...wishLists]

              for (let item of json) {
                const list = newWishLists.find(
                  ({ id }) => id === item.list_id
                )
                const index = list?.list_items?.findIndex(
                  ({ id }) => id === item.id
                )

                if (list && typeof index === 'number')
                  list.list_items[index] = item
              }

              setWishLists(newWishLists)
              onSuccess && onSuccess()
            } else {
              setFlashProps({
                hidden: false,
                type: 'error',
                message: UNEXPECTED_ERROR_MESSAGE,
              })

              onError && onError()
            }

            removeApiCall('wishListItems', 'patch')
          })
          .catch((e: ApiError) => {
            retries ??= 1

            if (e.code === 401 && retries > 0) {
              return withTokenRefresh((newToken) => {
                updateWishListItem(
                  itemId,
                  attributes,
                  onSuccess,
                  onError,
                  newToken,
                  (retries as number) - 1
                )
              })
            } else if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "You have attempted to update a wish list item that doesn't exist, or doesn't belong to you. Please refresh and try again.",
              })
            } else {
              handleApiError(e, 'list item')
            }

            removeApiCall('wishListItems', 'patch')
            onError && onError()
          })
      }
    },
    [token, wishLists]
  )

  /**
   *
   * Destroy a wish list item
   *
   */

  const destroyWishListItem = useCallback(
    (
      itemId: number,
      onSuccess?: CallbackFunction | null,
      onError?: CallbackFunction | null,
      idToken?: string | null,
      retries?: number
    ) => {
      idToken ??= token

      if (idToken) {
        addApiCall('wishListItems', 'delete')

        deleteWishListItem(itemId, idToken)
          .then(({ status, json }) => {
            if (status === 200) {
              const newWishLists = [...wishLists]

              newWishLists[0] = json[0]

              const index = newWishLists.findIndex(
                ({ id }) => id === json[1].id
              )
              newWishLists[index] = json[1]

              setWishLists(newWishLists)
              removeApiCall('wishListItems', 'delete')

              setFlashProps({
                hidden: false,
                type: 'success',
                message: 'Success! Your wish list item has been deleted.',
              })

              onSuccess && onSuccess()
            } else {
              removeApiCall('wishListItems', 'delete')

              setFlashProps({
                hidden: false,
                type: 'error',
                message: UNEXPECTED_ERROR_MESSAGE,
              })

              onError && onError()
            }
          })
          .catch((e: ApiError) => {
            retries ??= 1

            if (e.code === 401 && retries > 0) {
              return withTokenRefresh((newToken) => {
                destroyWishListItem(
                  itemId,
                  onSuccess,
                  onError,
                  newToken,
                  (retries as number) - 1
                )
              })
            } else if (e.code === 404) {
              setFlashProps({
                hidden: false,
                type: 'error',
                message:
                  "You have tried to delete a list item that doesn't exist, or doesn't belong to you. Please refresh and try again.",
              })
            } else {
              handleApiError(e, 'list item')
            }

            removeApiCall('wishListItems', 'delete')
            onError && onError()
          })
      }
    },
    [token, wishLists]
  )

  /**
   *
   * Set the context provider value
   *
   */

  const value = {
    wishLists,
    wishListsLoadingState,
    createWishList,
    updateWishList,
    destroyWishList,
    createWishListItem,
    updateWishListItem,
    destroyWishListItem,
  }

  /**
   *
   * Set the active game automatically from the query string
   * or, failing that, from the games themselves when they load
   *
   */

  useEffect(() => {
    const gameId: number = Number(queryString.get('gameId'))

    if (gameId > 0) {
      setActiveGame(gameId)
    } else if (gamesLoadingState === DONE && games.length) {
      setActiveGame(games[0].id)
    }
  }, [queryString, gamesLoadingState, games])

  useEffect(() => {
    if (authLoading) return

    // Only fetch wish lists if token is present and
    // (a) the token just changed from null to a string value or
    // (b) the token is already set and it is the initial render
    if (
      token &&
      (!previousTokenRef.current || previousTokenRef.current === token)
    )
      fetchWishLists()

    previousTokenRef.current = token
  }, [authLoading, activeGame, token])

  useEffect(() => {
    if (gamesLoadingState === DONE && !games.length)
      setWishListsLoadingState(DONE)
  }, [gamesLoadingState, games])

  useEffect(() => {
    requireLogin()
  }, [requireLogin])

  return (
    <WishListsContext.Provider value={value}>
      {children}
    </WishListsContext.Provider>
  )
}
