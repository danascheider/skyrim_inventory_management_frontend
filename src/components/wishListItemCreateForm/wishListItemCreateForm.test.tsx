import { describe, test, expect, vitest } from 'vitest'
import { act, fireEvent } from '@testing-library/react'
import { renderAuthenticated } from '../../support/testUtils'
import {
  gamesContextValue,
  wishListsContextValue,
} from '../../support/data/contextValues'
import { YELLOW } from '../../utils/colorSchemes'
import { ColorProvider } from '../../contexts/colorContext'
import { PageProvider } from '../../contexts/pageContext'
import { GamesContext } from '../../contexts/gamesContext'
import { WishListsContext } from '../../contexts/wishListsContext'
import WishListItemCreateForm from './wishListItemCreateForm'

describe('WishListItemCreateForm', () => {
  test('displays correct fields', () => {
    const wrapper = renderAuthenticated(
      <PageProvider>
        <GamesContext.Provider value={gamesContextValue}>
          <WishListsContext.Provider value={wishListsContextValue}>
            <ColorProvider colorScheme={YELLOW}>
              <WishListItemCreateForm listId={4} />
            </ColorProvider>
          </WishListsContext.Provider>
        </GamesContext.Provider>
      </PageProvider>
    )

    expect(wrapper.getByText('Add item to list...')).toBeTruthy()
    expect(wrapper.getByLabelText('Description')).toBeTruthy()
    expect(wrapper.getByLabelText('Quantity')).toBeTruthy()
    expect(wrapper.getByLabelText('Unit Weight')).toBeTruthy()
    expect(wrapper.getByLabelText('Notes')).toBeTruthy()
    expect(wrapper.getByText('Add to List')).toBeTruthy()
  })

  test('matches snapshot', () => {
    const wrapper = renderAuthenticated(
      <PageProvider>
        <GamesContext.Provider value={gamesContextValue}>
          <WishListsContext.Provider value={wishListsContextValue}>
            <ColorProvider colorScheme={YELLOW}>
              <WishListItemCreateForm listId={4} />
            </ColorProvider>
          </WishListsContext.Provider>
        </GamesContext.Provider>
      </PageProvider>
    )

    expect(wrapper).toMatchSnapshot()
  })

  describe('submitting the form', () => {
    test('submits the form when attributes are valid', () => {
      const createWishListItem = vitest.fn()

      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <WishListsContext.Provider
              value={{ ...wishListsContextValue, createWishListItem }}
            >
              <ColorProvider colorScheme={YELLOW}>
                <WishListItemCreateForm listId={4} />
              </ColorProvider>
            </WishListsContext.Provider>
          </GamesContext.Provider>
        </PageProvider>
      )

      const descField = wrapper.getByLabelText('Description')
      const quantityField = wrapper.getByLabelText('Quantity')
      const form = wrapper.getByLabelText('Wish list item creation form')

      fireEvent.change(descField, { target: { value: '  Iron ingot    ' } })
      fireEvent.change(quantityField, { target: { value: '2' } })

      act(() => fireEvent.submit(form))

      expect(createWishListItem).toHaveBeenCalledWith(
        4,
        {
          description: 'Iron ingot',
          quantity: 2,
        },
        expect.any(Function),
        expect.any(Function)
      )
    })

    test("doesn't submit the form if there are invalid values", () => {
      const createWishListItem = vitest.fn()

      const wrapper = renderAuthenticated(
        <PageProvider>
          <GamesContext.Provider value={gamesContextValue}>
            <WishListsContext.Provider
              value={{ ...wishListsContextValue, createWishListItem }}
            >
              <ColorProvider colorScheme={YELLOW}>
                <WishListItemCreateForm listId={4} />
              </ColorProvider>
            </WishListsContext.Provider>
          </GamesContext.Provider>
        </PageProvider>
      )

      const descField = wrapper.getByLabelText('Description')
      const quantityField = wrapper.getByLabelText('Quantity')

      // In this test, we have to click the button to simulate the form being
      // submitted by a user. Firing the `submit` event directly on the form
      // effectively takes as a foregone conclusion that the form has submitted
      // successfully - validations will be skipped.
      const button = wrapper.getByText('Add to List')

      act(() => {
        fireEvent.change(descField, { target: { value: '' } })
        fireEvent.change(quantityField, { target: { value: '2' } })
        fireEvent.click(button)
      })

      expect(createWishListItem).not.toHaveBeenCalled()
    })
  })
})
