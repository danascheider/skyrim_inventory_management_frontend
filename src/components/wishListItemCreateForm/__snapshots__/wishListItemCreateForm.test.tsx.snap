// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShoppingListItemCreateForm > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_root_8112da _collapsed_8112da"
        style="--base-color: #ffcb32; --hover-color: #ffc519; --body-color: #fff2cc; --border-color: #cc9800; --text-color-main: #4c3900; --text-color-secondary: #7f5700;"
      >
        <div
          class="_triggerContainer_8112da"
        >
          <button
            aria-controls="addItemToShoppingList4"
            aria-expanded="false"
            class="_trigger_8112da"
          >
            Add item to list...
          </button>
        </div>
        <div
          aria-hidden="true"
          class="rah-static rah-static--height-zero "
          id="addItemToShoppingList4"
          style="height: 0px; overflow: hidden;"
        >
          <div
            style="display: none;"
          >
            <form
              aria-label="Wish list item creation form"
              class="_form_8112da"
            >
              <label
                class="_label_8112da"
              >
                Description
                <input
                  class="_input_8112da"
                  name="description"
                  placeholder="Description"
                  required=""
                  type="text"
                />
              </label>
              <label
                class="_label_8112da"
              >
                Quantity
                <input
                  class="_input_8112da"
                  inputmode="numeric"
                  min="1"
                  name="quantity"
                  pattern="^[0-9]*$"
                  placeholder="Quantity"
                  required=""
                  step="1"
                  type="number"
                  value="1"
                />
              </label>
              <label
                class="_label_8112da"
              >
                Unit Weight
                <input
                  class="_input_8112da"
                  inputmode="decimal"
                  min="0"
                  name="unit_weight"
                  placeholder="Unit Weight"
                  step="0.1"
                  type="number"
                />
              </label>
              <label
                class="_label_8112da"
              >
                Notes
                <input
                  class="_input_8112da"
                  name="notes"
                  placeholder="Notes"
                  type="text"
                />
              </label>
              <button
                class="_submit_8112da"
                type="submit"
              >
                Add to List
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_root_8112da _collapsed_8112da"
      style="--base-color: #ffcb32; --hover-color: #ffc519; --body-color: #fff2cc; --border-color: #cc9800; --text-color-main: #4c3900; --text-color-secondary: #7f5700;"
    >
      <div
        class="_triggerContainer_8112da"
      >
        <button
          aria-controls="addItemToShoppingList4"
          aria-expanded="false"
          class="_trigger_8112da"
        >
          Add item to list...
        </button>
      </div>
      <div
        aria-hidden="true"
        class="rah-static rah-static--height-zero "
        id="addItemToShoppingList4"
        style="height: 0px; overflow: hidden;"
      >
        <div
          style="display: none;"
        >
          <form
            aria-label="Wish list item creation form"
            class="_form_8112da"
          >
            <label
              class="_label_8112da"
            >
              Description
              <input
                class="_input_8112da"
                name="description"
                placeholder="Description"
                required=""
                type="text"
              />
            </label>
            <label
              class="_label_8112da"
            >
              Quantity
              <input
                class="_input_8112da"
                inputmode="numeric"
                min="1"
                name="quantity"
                pattern="^[0-9]*$"
                placeholder="Quantity"
                required=""
                step="1"
                type="number"
                value="1"
              />
            </label>
            <label
              class="_label_8112da"
            >
              Unit Weight
              <input
                class="_input_8112da"
                inputmode="decimal"
                min="0"
                name="unit_weight"
                placeholder="Unit Weight"
                step="0.1"
                type="number"
              />
            </label>
            <label
              class="_label_8112da"
            >
              Notes
              <input
                class="_input_8112da"
                name="notes"
                placeholder="Notes"
                type="text"
              />
            </label>
            <button
              class="_submit_8112da"
              type="submit"
            >
              Add to List
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
